// node-appwrite/dist/index.js
'use strict';

var client = require('./client');
var account = require('./services/account');
var avatars = require('./services/avatars');
var databases = require('./services/databases');
var functions = require('./services/functions');
var graphql = require('./services/graphql');
var health = require('./services/health');
var locale = require('./services/locale');
var messaging = require('./services/messaging');
var sites = require('./services/sites');
var storage = require('./services/storage');
var tablesDb = require('./services/tables-db');
var teams = require('./services/teams');
var tokens = require('./services/tokens');
var users = require('./services/users');
var permission = require('./permission');
var role = require('./role');
var id = require('./id');
var operator = require('./operator');
var authenticatorType = require('./enums/authenticator-type');
var authenticationFactor = require('./enums/authentication-factor');
var oAuthProvider = require('./enums/o-auth-provider');
var browser = require('./enums/browser');
var creditCard = require('./enums/credit-card');
var flag = require('./enums/flag');
var relationshipType = require('./enums/relationship-type');
var relationMutate = require('./enums/relation-mutate');
var indexType = require('./enums/index-type');
var runtime = require('./enums/runtime');
var vcsDeploymentType = require('./enums/vcs-deployment-type');
var deploymentDownloadType = require('./enums/deployment-download-type');
var executionMethod = require('./enums/execution-method');
var name = require('./enums/name');
var messagePriority = require('./enums/message-priority');
var smtpEncryption = require('./enums/smtp-encryption');
var framework = require('./enums/framework');
var buildRuntime = require('./enums/build-runtime');
var adapter = require('./enums/adapter');
var compression = require('./enums/compression');
var imageGravity = require('./enums/image-gravity');
var imageFormat = require('./enums/image-format');
var passwordHash = require('./enums/password-hash');
var messagingProviderType = require('./enums/messaging-provider-type');
var databaseType = require('./enums/database-type');
var attributeStatus = require('./enums/attribute-status');
var columnStatus = require('./enums/column-status');
var indexStatus = require('./enums/index-status');
var deploymentStatus = require('./enums/deployment-status');
var executionTrigger = require('./enums/execution-trigger');
var executionStatus = require('./enums/execution-status');
var healthAntivirusStatus = require('./enums/health-antivirus-status');
var healthCheckStatus = require('./enums/health-check-status');
var messageStatus = require('./enums/message-status');



Object.defineProperty(exports, 'AppwriteException', {
  enumerable: true,
  get: function () { return client.AppwriteException; }
});
Object.defineProperty(exports, 'Client', {
  enumerable: true,
  get: function () { return client.Client; }
});
Object.defineProperty(exports, 'Query', {
  enumerable: true,
  get: function () { return client.Query; }
});
Object.defineProperty(exports, 'Account', {
  enumerable: true,
  get: function () { return account.Account; }
});
Object.defineProperty(exports, 'Avatars', {
  enumerable: true,
  get: function () { return avatars.Avatars; }
});
Object.defineProperty(exports, 'Databases', {
  enumerable: true,
  get: function () { return databases.Databases; }
});
Object.defineProperty(exports, 'Functions', {
  enumerable: true,
  get: function () { return functions.Functions; }
});
Object.defineProperty(exports, 'Graphql', {
  enumerable: true,
  get: function () { return graphql.Graphql; }
});
Object.defineProperty(exports, 'Health', {
  enumerable: true,
  get: function () { return health.Health; }
});
Object.defineProperty(exports, 'Locale', {
  enumerable: true,
  get: function () { return locale.Locale; }
});
Object.defineProperty(exports, 'Messaging', {
  enumerable: true,
  get: function () { return messaging.Messaging; }
});
Object.defineProperty(exports, 'Sites', {
  enumerable: true,
  get: function () { return sites.Sites; }
});
Object.defineProperty(exports, 'Storage', {
  enumerable: true,
  get: function () { return storage.Storage; }
});
Object.defineProperty(exports, 'TablesDB', {
  enumerable: true,
  get: function () { return tablesDb.TablesDB; }
});
Object.defineProperty(exports, 'Teams', {
  enumerable: true,
  get: function () { return teams.Teams; }
});
Object.defineProperty(exports, 'Tokens', {
  enumerable: true,
  get: function () { return tokens.Tokens; }
});
Object.defineProperty(exports, 'Users', {
  enumerable: true,
  get: function () { return users.Users; }
});
Object.defineProperty(exports, 'Permission', {
  enumerable: true,
  get: function () { return permission.Permission; }
});
Object.defineProperty(exports, 'Role', {
  enumerable: true,
  get: function () { return role.Role; }
});
Object.defineProperty(exports, 'ID', {
  enumerable: true,
  get: function () { return id.ID; }
});
Object.defineProperty(exports, 'Condition', {
  enumerable: true,
  get: function () { return operator.Condition; }
});
Object.defineProperty(exports, 'Operator', {
  enumerable: true,
  get: function () { return operator.Operator; }
});
Object.defineProperty(exports, 'AuthenticatorType', {
  enumerable: true,
  get: function () { return authenticatorType.AuthenticatorType; }
});
Object.defineProperty(exports, 'AuthenticationFactor', {
  enumerable: true,
  get: function () { return authenticationFactor.AuthenticationFactor; }
});
Object.defineProperty(exports, 'OAuthProvider', {
  enumerable: true,
  get: function () { return oAuthProvider.OAuthProvider; }
});
Object.defineProperty(exports, 'Browser', {
  enumerable: true,
  get: function () { return browser.Browser; }
});
Object.defineProperty(exports, 'CreditCard', {
  enumerable: true,
  get: function () { return creditCard.CreditCard; }
});
Object.defineProperty(exports, 'Flag', {
  enumerable: true,
  get: function () { return flag.Flag; }
});
Object.defineProperty(exports, 'RelationshipType', {
  enumerable: true,
  get: function () { return relationshipType.RelationshipType; }
});
Object.defineProperty(exports, 'RelationMutate', {
  enumerable: true,
  get: function () { return relationMutate.RelationMutate; }
});
Object.defineProperty(exports, 'IndexType', {
  enumerable: true,
  get: function () { return indexType.IndexType; }
});
Object.defineProperty(exports, 'Runtime', {
  enumerable: true,
  get: function () { return runtime.Runtime; }
});
Object.defineProperty(exports, 'VCSDeploymentType', {
  enumerable: true,
  get: function () { return vcsDeploymentType.VCSDeploymentType; }
});
Object.defineProperty(exports, 'DeploymentDownloadType', {
  enumerable: true,
  get: function () { return deploymentDownloadType.DeploymentDownloadType; }
});
Object.defineProperty(exports, 'ExecutionMethod', {
  enumerable: true,
  get: function () { return executionMethod.ExecutionMethod; }
});
Object.defineProperty(exports, 'Name', {
  enumerable: true,
  get: function () { return name.Name; }
});
Object.defineProperty(exports, 'MessagePriority', {
  enumerable: true,
  get: function () { return messagePriority.MessagePriority; }
});
Object.defineProperty(exports, 'SmtpEncryption', {
  enumerable: true,
  get: function () { return smtpEncryption.SmtpEncryption; }
});
Object.defineProperty(exports, 'Framework', {
  enumerable: true,
  get: function () { return framework.Framework; }
});
Object.defineProperty(exports, 'BuildRuntime', {
  enumerable: true,
  get: function () { return buildRuntime.BuildRuntime; }
});
Object.defineProperty(exports, 'Adapter', {
  enumerable: true,
  get: function () { return adapter.Adapter; }
});
Object.defineProperty(exports, 'Compression', {
  enumerable: true,
  get: function () { return compression.Compression; }
});
Object.defineProperty(exports, 'ImageGravity', {
  enumerable: true,
  get: function () { return imageGravity.ImageGravity; }
});
Object.defineProperty(exports, 'ImageFormat', {
  enumerable: true,
  get: function () { return imageFormat.ImageFormat; }
});
Object.defineProperty(exports, 'PasswordHash', {
  enumerable: true,
  get: function () { return passwordHash.PasswordHash; }
});
Object.defineProperty(exports, 'MessagingProviderType', {
  enumerable: true,
  get: function () { return messagingProviderType.MessagingProviderType; }
});
Object.defineProperty(exports, 'DatabaseType', {
  enumerable: true,
  get: function () { return databaseType.DatabaseType; }
});
Object.defineProperty(exports, 'AttributeStatus', {
  enumerable: true,
  get: function () { return attributeStatus.AttributeStatus; }
});
Object.defineProperty(exports, 'ColumnStatus', {
  enumerable: true,
  get: function () { return columnStatus.ColumnStatus; }
});
Object.defineProperty(exports, 'IndexStatus', {
  enumerable: true,
  get: function () { return indexStatus.IndexStatus; }
});
Object.defineProperty(exports, 'DeploymentStatus', {
  enumerable: true,
  get: function () { return deploymentStatus.DeploymentStatus; }
});
Object.defineProperty(exports, 'ExecutionTrigger', {
  enumerable: true,
  get: function () { return executionTrigger.ExecutionTrigger; }
});
Object.defineProperty(exports, 'ExecutionStatus', {
  enumerable: true,
  get: function () { return executionStatus.ExecutionStatus; }
});
Object.defineProperty(exports, 'HealthAntivirusStatus', {
  enumerable: true,
  get: function () { return healthAntivirusStatus.HealthAntivirusStatus; }
});
Object.defineProperty(exports, 'HealthCheckStatus', {
  enumerable: true,
  get: function () { return healthCheckStatus.HealthCheckStatus; }
});
Object.defineProperty(exports, 'MessageStatus', {
  enumerable: true,
  get: function () { return messageStatus.MessageStatus; }
});
